packages/types/src/errors/tfchain/index.ts

```diff
@@ -1,6 +1,155 @@
-export * as dao from "./dao";
-export * as tfkvStore from "./kvstore";
-export * as smartContractModule from "./smart_contract";
-export * as tfgridModule from "./tfgrid";
-export * as tftBridgeModule from "./tft_bridge";
-export * as tftPriceModule from "./tft_price";
+import { ApiPromise } from "@polkadot/api";
+import { SubmittableExtrinsic } from "@polkadot/api/types";
+import type { DispatchError } from "@polkadot/types/interfaces";
+import { ISubmittableResult } from "@polkadot/types/types";
+import { AnyTuple } from "@polkadot/types-codec/types";
+
+import { InsufficientBalanceError } from "../base_error";
+
+interface ITFChainError {
+  message: string;
+  keyError?: string;
+  section?: string;
+  method?: string;
+  args?: AnyTuple;
+  docs?: string[];
+}
+
+class TFChainError extends Error implements ITFChainError {
+  args?: AnyTuple;
+  keyError?: string;
+  section?: string;
+  method?: string;
+  docs?: string[];
+
+  constructor(options: ITFChainError) {
+    super(options.message);
+    this.name = "TFChainError";
+    this.keyError = options.keyError;
+    this.section = options.section;
+    this.args = options.args;
+    this.method = options.method;
+    this.docs = options.docs;
+  }
+}
+
+class TFChainErrorWrapper {
+  error: DispatchError | Error;
+  extrinsic: SubmittableExtrinsic<"promise", ISubmittableResult>;
+  api: ApiPromise;
+
+  constructor(
+    error: DispatchError | Error,
+    extrinsic: SubmittableExtrinsic<"promise", ISubmittableResult>,
+    api: ApiPromise,
+  ) {
+    this.error = error;
+    this.extrinsic = extrinsic;
+    this.api = api;
+  }
+
+  /**
+   * Throws a TFChainError based on the type of error encountered:
+   * - If the error is a dispatch error:
+   *   - If it is a module error, throws a TFChainError with module details.
+   *   - If it is a token error, throws a TFChainError with token details.
+   *   - If it is an arithmetic error, throws a TFChainError with arithmetic details.
+   *   - If it is a bad origin error, throws a TFChainError for bad origin.
+   *   - If it is an unknown dispatch error, throws a TFChainError for unknown dispatch error.
+   * - If the error is an instance of Error:
+   *   - If the message includes "Inability to pay some fees", throws an InsufficientBalanceError.
+   *   - Otherwise, throws a TFChainError for a generic error.
+   * - If the error is not a dispatch error or an instance of Error, throws a TFChainError for an unknown error.
+   * @returns {TFChainError} The specific TFChainError thrown based on the encountered error type.
+   */
+  throw(): TFChainError {
+    const args = this.extrinsic.method.args;
+    const method = this.extrinsic.method.method;
```

Comment:
Please don't use the extrinsic data, it may invalid error data https://github.com/threefoldtech/tfgrid-sdk-ts/issues/2896, it happened before and we decided to get the exact error method https://github.com/threefoldtech/tfgrid-sdk-ts/pull/2995

we can't rely on the extrinsic